boskos/metrics/metrics.go

```diff
@@ -90,51 +67,53 @@ func main() {
 		states = defaultStates
 	}
 
-	initMetrics()
-
-	http.Handle("/prometheus", promhttp.Handler())
-	http.Handle("/", handleMetric(boskos))
+	metrics.ExposeMetrics("boskos", config.PushGateway{})
 
 	go func() {
-		logTick := time.NewTicker(time.Minute).C
+		logTick := time.NewTicker(30 * time.Second).C
 		for range logTick {
 			if err := update(boskos); err != nil {
-				logrus.WithError(err).Warning("[Boskos Metrics]Update failed!")
+				logrus.WithError(err).Warning("Update failed!")
 			}
 		}
 	}()
 
 	logrus.Info("Start Service")
-	logrus.WithError(http.ListenAndServe(":8080", nil)).Fatal("ListenAndServe returned.")
+	metricsMux := http.NewServeMux()
+	metricsMux.Handle("/", handleMetric(boskos))
+	logrus.WithError(http.ListenAndServe(":8080", metricsMux)).Fatal("ListenAndServe returned.")
 }
 
-func filterMetrics(src map[string]int) map[string]int {
-	metricStates := sets.NewString(states...)
-	dest := map[string]int{}
-	// Making sure all metrics are created
-	for state := range metricStates {
-		dest[state] = 0
-	}
-	dest[common.Other] = 0
-	for state, value := range src {
-		if state != common.Other && metricStates.Has(state) {
-			dest[state] = value
-		} else {
-			dest[common.Other] += value
+func update(boskos *client.Client) error {
+	// initialize resources counted by type, then state
+	resourcesByState := map[string]map[string]float64{}
+	for _, resource := range resources {
+		resourcesByState[resource] = map[string]float64{}
+		for _, state := range states {
+			resourcesByState[resource][state] = 0
```

Comment:
> The current behavior
of producing a "0" count for type/state combinations that do not have
any record in Boskos but are requested via flag is preserved.

I guess that this line is related to the above. But I cannot see why we need `0` here as init value.

Reply:
we still want to be able to query empty state, without initialize I think there will be no value, instead of 0?

Reply:
Maybe we can deal with no value by the query `exression OR on() vector(0)`
https://github.com/grafana/grafana/issues/2393#issuecomment-192522042

Reply:
Hmmm @hongkailiu so the case that I think @krzyzacy was worried about is -- if we have a resource `myThing` that has some state `cleaning` but only transitively goes into that state -- say, from `inUse` --> `dirty` --> `cleaning` --> `clean`. If we do a query like `sum(boskos_resources{type="myThing"}) by (state)` we will get no metric for states like `cleaning` when `boskos` has no resources in that state, instead of `0`. I don't see how to use your suggestion for that case, since we aren't calling out states individually.

Reply:
I see now. Need to fix the test?
/lgtm